src/QsFmt/Formatter/SyntaxTree/Statement.fs

```diff
@@ -25,7 +56,43 @@ type Return =
         Semicolon: Terminal
     }
 
-type If =
+type Use =
+    {
+        UseKeyword: Terminal
+        Binding: QubitBinding
+        OpenParen: Terminal option
+        CloseParen: Terminal option
+        Semicolon: Terminal
+    }
+
+type Borrow =
+    {
+        BorrowKeyword: Terminal
+        Binding: QubitBinding
+        OpenParen: Terminal option
+        CloseParen: Terminal option
+        Semicolon: Terminal
+    }
+
+type UseBlock =
+    {
+        UseKeyword: Terminal
+        Binding: QubitBinding
+        OpenParen: Terminal option
+        CloseParen: Terminal option
+        Block: Statement Block
+    }
+
+and BorrowBlock =
+    {
+        BorrowKeyword: Terminal
+        Binding: QubitBinding
+        OpenParen: Terminal option
+        CloseParen: Terminal option
+        Block: Statement Block
+    }
```

Comment:
I will look into this, but I need to ask a broader question. If we make it easier to write transformation that don't need to know about the distinction, we make it harder to write transformations that _do_ need to make the distinction. How do we strike the appropriate balance and tell which direction to go in and which side to prefer?

Reply:
> If we make it easier to write transformation that don't need to know about the distinction, we make it harder to write transformations that *do* need to make the distinction.

Hmm, I don't think I agree with this. It's easier to filter data of the same type than it is to combine data of different types. For example, if you had a node like this:

```fsharp
type QubitBindingKeyword = Use of Terminal | Borrow of Terminal
type QubitBinding = { BindingKeyword: QubitBindingKeyword; ... }
```

Then you could write a transformation like this:

```fsharp
member _.QubitBinding binding =
    match binding.BindingKeyword with
    | Use _ -> failwith "in a use block"
    | Borrow _ -> failwith "in a borrow block"
```

Reply:
In the sense that the transformation handles document how transformations should be written and guide developers thinking on how transformations are written, not having separate handles for something that is distinguishable means that that distinction will be more hidden and transformations will be written biased towards not making that distinction. In the example you provided, it will feel more unnatural/unsupported to write a transformation that makes a distinction between Use and Borrow than if there were separate handles for Use and Borrow. The code would seem to suggest, by not having two separate handles, that you as a developer are not supposed to write transformations that make a distinction between the two.

Reply:
Sorry my response seemed like that used a lot of abstract language. Did it make sense, or should I try to reword it?

Reply:
> The code would seem to suggest, by not having two separate handles, that you as a developer are not supposed to write transformations that make a distinction between the two.

No, the code doesn't suggest that. There are an infinite number of ways to divide up nodes in the syntax tree. For example, you could select all empty blocks, or all blocks that have exactly one statement in them. Or you could select all qubit bindings that allocate an array of qubits. It's not practical to write separate handlers for all of these cases. It's perfectly natural and encouraged to take the existing handles and branch on the values to do more specific things.

You are also conflating two different things: (1) the representation of the nodes in the syntax tree, and (2) the handlers provided by the transformation classes. Those are not the same. For example, the compiler's syntax tree uses one node for both use and borrow:

https://github.com/microsoft/qsharp-compiler/blob/6d20b62f8822044974e3ae0027df910158454bd4/src/QsCompiler/DataStructures/SyntaxTree.fs#L854-L855

But it provides three transformation handlers (general qubit scope, use scope, and borrow scope):

https://github.com/microsoft/qsharp-compiler/blob/6d20b62f8822044974e3ae0027df910158454bd4/src/QsCompiler/Core/Transformations/StatementTransformation.fs#L195-L206

Reply:
I agree that there are numerous ways to structure the data types for the syntax tree. But then how do we decided to do one way versus the others? What's the criteria that suggests we should have the same node for use and borrow blocks instead of different ones? And I think that, while the handles and data types are different, they should be one-to-one when reasonable.

Reply:
The criteria I'm using is when two types share substantial common factors, those factors should often be pulled out. That allows them to be reasoned about as smaller individual pieces. The larger piece (e.g. a use statement with a block) can be created by composing the smaller pieces back together.

Similar reasoning applies to refactoring functions: several smaller functions that are composed together are easier to read and reason about than one giant function that does everything.

I guess an analogy is factoring a quadratic. It's easy to look at (x + 2)(x - 4) and (x + 2)(x + 3) and see that they both share a common factor of x + 2, implying they are both zero at x = -2. It's a lot harder to notice that with the equivalent expressions x^2 - 2x - 8 and x^2 + 5x + 6. And it's a lot easier to go from factored to un-factored than the other direction.

Reply:
This is roughly my idea:

```fsharp
type QubitDeclarationKind =
    | Use
    | Borrow

type QubitDeclarationCoda =
    | Semicolon of Terminal
    | Block of Statement Block

type QubitDeclaration =
    {
        Kind: QubitDeclarationKind
        Keyword: Terminal
        OpenParen: Terminal option
        Binding: QubitBinding
        CloseParen: Terminal option
        Coda: QubitDeclarationCoda
    }
```

Reply:
This is a very nice a clean implementation. I've added the changes for it and I think it worked great! Thanks Sarah.